### PR TITLE
Bootstrapping the Terraform Provider

### PR DESCRIPTION
**Pull request by graham-russell**
- Starting work on creating the ably provider code. No resources implemented yet. Follow up PRs/work to do this.
- Setting up the repo based on https://github.com/hashicorp/terraform-provider-scaffolding-framework
- Setting up documentation using tfplugindocs
- Provide some working docs for local development
- Moved terraform files from examples to examples/resources to keep inline with provider conventions

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+
+
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `token` (String, Sensitive)
+- `url` (String)
```
Comment (kavalerov) [inline docs/index.md:21]: Should URL be optional? Or maybe we can replace / add alternative parameter `version`?

File: internal/provider/provider.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		token = os.Getenv("ABLY_ACCOUNT_TOKEN")
+	} else {
+		token = config.Token.Value
+	}
+
+	if token == "" {
+		// Error vs warning - empty value must stop execution
+		resp.Diagnostics.AddError(
+			"Unable to find Ably token",
+			"Username cannot be an empty string",
```
Comment (kavalerov) [inline internal/provider/provider.go:75]: I am not sure what "username" is referred to here.

File: internal/provider/provider.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		return
+	}
+
+	// User must specify an Ably Control API Url
+	var url string
+	if config.Url.Unknown {
+		// Cannot connect to client with an unknown value
+		resp.Diagnostics.AddError(
+			"Unable to create client",
+			"Cannot use unknown value as Urlt",
```
Comment (kavalerov) [inline internal/provider/provider.go:86]: Should this be something like "Ably Control API URL is required"?

File: internal/provider/provider.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		url = os.Getenv("ABLY_URL")
+	} else {
+		url = config.Url.Value
+	}
+
+	if url == "" {
+		// Error vs warning - empty value must stop execution
+		resp.Diagnostics.AddError(
+			"Unable to find url",
+			"Url cannot be an empty string",
```
Comment (kavalerov) [inline internal/provider/provider.go:101]: Same as comment above - should this be "Ably Control API URL is required"?

File: internal/provider/provider.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	configured bool
+	client     *ably_control_go.Client
+}
+
+// GetSchema
+func (p *provider) GetSchema(_ context.Context) (tfsdk.Schema, diag.Diagnostics) {
+	return tfsdk.Schema{
+		Attributes: map[string]tfsdk.Attribute{
+			"token": {
+				Type:     types.StringType,
```
Comment (kavalerov) [inline internal/provider/provider.go:27]: Very minor, but indentation looks off to me, between this line and following lines 😅
Please ignore this comment if it is the way it should be in Go, it just may be my lack of experience with the language.

**Review by kavalerov — CHANGES_REQUESTED** [review]

File: docs/index.md (hunk 1 above)
Comment (graham-russell) [reply docs/index.md:21]: What would the version parameter be used for?

I'm thinking perhaps we can use a default url of "https://control.ably.net/v1". 
For testing, I'm using our staging endpoint with pre-prod credentials.

File: internal/provider/provider.go (hunk 5 above)
Comment (graham-russell) [reply internal/provider/provider.go]: Fixed

File: internal/provider/provider.go (hunk 4 above)
Comment (graham-russell) [reply internal/provider/provider.go]: Fixed

File: internal/provider/provider.go (hunk 3 above)
Comment (graham-russell) [reply internal/provider/provider.go]: Fixed

File: internal/provider/provider.go (hunk 2 above)
Comment (graham-russell) [reply internal/provider/provider.go]: Fixed

File: docs/index.md (hunk 1 above)
Comment (graham-russell) [reply docs/index.md:21]: I've made the url parameter optional now. In the provider, if url == "", we use a default of "https://control.ably.net/v1".

**Review by kavalerov — APPROVED** [review]
LGTM, thanks

**Pull request merged**